File: backend/src/modules/pedido/pedido.controller.ts
import {
  Controller,
  Get,
  Post,
  Put,
  Body,
  Param,
  Query,
  HttpCode,
  HttpStatus,
  UseGuards,
} from '@nestjs/common';
import { ApiOperation, ApiResponse, ApiTags, ApiBearerAuth } from '@nestjs/swagger';
import { Response } from 'express';
import { Res } from '@nestjs/common';
import { PedidoService, CreatePedidoResult } from './pedido.service';
import { CreatePedidoDto } from './dto/create-pedido.dto';
import { PedidoInstruccionesPagoDto } from './dto/pedido-instrucciones-pago.dto';
import { PedidoPendienteDto } from './dto/pedido-pendiente.dto';
import { CancelarPedidoResponse } from './dto/cancelar-pedido-response.dto';
import { ExportarPedidosQueryDto } from './dto/exportar-pedidos-query.dto';
import { JwtAuthGuard } from '@modules/auth/guards/jwt-auth.guard';

@ApiTags('pedidos')
@Controller('pedidos')
export class PedidoController {
  constructor(private readonly pedidoService: PedidoService) {}

  @UseGuards(JwtAuthGuard)
  @ApiBearerAuth()
  @Get('list-pendientes')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Listar pedidos pendientes de pago (admin)' })
  @ApiResponse({ status: 200, description: 'Pedidos pendientes obtenidos' })
  async listarPendientes(
    @Query('pagina') pagina?: number,
    @Query('tamano') tamano?: number,
  ) {
    const filtros: { pagina?: number; tamano?: number } = {};
    if (pagina != null) filtros.pagina = pagina;
    if (tamano != null) filtros.tamano = tamano;
    return this.pedidoService.listarPendientes(filtros);
  }

  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ summary: 'Crear un nuevo pedido (público)' })
  @ApiResponse({ status: 201, description: 'Pedido creado exitosamente' })
  @ApiResponse({ status: 400, description: 'Error de validación' })
  async crear(@Body() dto: CreatePedidoDto): Promise<{ mensaje: string; pedido: CreatePedidoResult }> {
    return this.pedidoService.crear(dto);
  }

  @Get('buscar')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Buscar pedido por codigo y email (publico)' })
  @ApiResponse({ status: 200, description: 'Pedido encontrado' })
  @ApiResponse({ status: 404, description: 'No se encontro un pedido con esos datos' })
  async buscarPorCodigoYEmail(
    @Query('codigo') codigo: string,
    @Query('email') email: string,
  ) {
    return this.pedidoService.buscarPorCodigoYEmail(codigo, email);
  }

  @UseGuards(JwtAuthGuard)
  @ApiBearerAuth()
  @Get('export')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Exportar pedidos a CSV (admin)' })
  @ApiResponse({ status: 200, description: 'Archivo CSV con todos los pedidos' })
  @ApiResponse({ status: 401, description: 'No autenticado' })
  async exportarCsv(
    @Query() query: ExportarPedidosQueryDto,
    @Res({ passthrough: true }) res: Response,
  ): Promise<string> {
    const pedidos = await this.pedidoService.listarTodosParaExport(query);
    const csv = PedidoService.generarCsv(pedidos, { agregarBOM: true });

    const fecha = new Date().toISOString().slice(0, 10);
    res.setHeader('Content-Type', 'text/csv; charset=utf-8');
    res.setHeader(
      'Content-Disposition',
      `attachment; filename="pedidos-${fecha}.csv"`,
    );
    return csv;
  }

  @Get(':id')
  @ApiOperation({ summary: 'Obtener un pedido por ID (público)' })
  @ApiResponse({ status: 200, description: 'Pedido encontrado' })
  @ApiResponse({ status: 404, description: 'Pedido no encontrado' })
  async obtenerUno(@Param('id') id: string) {
    return this.pedidoService.obtenerUno(id);
  }

  @Get(':id/instrucciones-pago')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Obtener instrucciones de pago para un pedido' })
  @ApiResponse({ status: 200, description: 'Instrucciones de pago obtenidas' })
  @ApiResponse({ status: 404, description: 'Pedido no encontrado' })
  @ApiResponse({ status: 503, description: 'Datos de pago no disponibles' })
  async obtenerInstruccionesPago(@Param('id') pedidoId: string): Promise<PedidoInstruccionesPagoDto> {
    return this.pedidoService.obtenerInstruccionesPago(pedidoId);
  }

  @UseGuards(JwtAuthGuard)
  @ApiBearerAuth()
  @Put(':id/confirmar-pago')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Confirmar pago manualmente de un pedido (admin)' })
  @ApiResponse({ status: 200, description: 'Pago confirmado exitosamente' })
  @ApiResponse({ status: 400, description: 'Pedido ya confirmado o stock insuficiente' })
  @ApiResponse({ status: 404, description: 'Pedido o producto no encontrado' })
  async confirmarPago(@Param('id') id: string) {
    return this.pedidoService.confirmarPago(id);
  }

  @UseGuards(JwtAuthGuard)
  @ApiBearerAuth()
  @Put(':id/cancelar')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Cancelar un pedido pendiente (admin)' })
  @ApiResponse({ status: 200, description: 'Pedido cancelado exitosamente' })
  @ApiResponse({ status: 400, description: 'No se puede cancelar un pedido confirmado o ya cancelado' })
  @ApiResponse({ status: 404, description: 'Pedido no encontrado' })
  async cancelar(@Param('id') id: string): Promise<CancelarPedidoResponse> {
    return this.pedidoService.cancelar(id);
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/pedido/PedidoController.java
package com.sistemapedidos.modules.pedido;

import com.sistemapedidos.domain.entity.Pedido;
import com.sistemapedidos.modules.pedido.dto.CreatePedidoDto;
import com.sistemapedidos.modules.pedido.dto.ExportarPedidosQueryDto;
import com.sistemapedidos.modules.pedido.dto.PedidoDetalle;
import com.sistemapedidos.modules.pedido.dto.PedidoInstruccionesPagoDto;
import com.sistemapedidos.modules.pedido.dto.PedidoOperacionesDto;
import jakarta.servlet.http.HttpServletResponse;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;
import java.time.LocalDate;
import java.util.List;

@RestController
@RequestMapping("/pedidos")
@RequiredArgsConstructor
public class PedidoController {

    private final PedidoService pedidoService;

    @GetMapping("/list-pendientes")
    public PedidoOperacionesDto.PaginadoPedidosPendientes listarPendientes(
            @RequestParam(defaultValue = "1") int pagina,
            @RequestParam(defaultValue = "20") int tamano) {
        return pedidoService.listarPendientes(pagina, tamano);
    }

    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    public PedidoOperacionesDto.CreatePedidoResponse crear(@Valid @RequestBody CreatePedidoDto dto) {
        return pedidoService.crear(dto);
    }

    @GetMapping("/buscar")
    public PedidoDetalle buscarPorCodigoYEmail(@RequestParam String codigo, @RequestParam String email) {
        return pedidoService.buscarPorCodigoYEmail(codigo, email);
    }

    @GetMapping("/export")
    public void exportarCsv(ExportarPedidosQueryDto query, HttpServletResponse response) throws IOException {
        List<Pedido> pedidos = pedidoService.listarTodosParaExport(query);
        String csv = PedidoService.generarCsv(pedidos);

        response.setContentType("text/csv; charset=utf-8");
        response.setCharacterEncoding("UTF-8");
        response.setHeader("Content-Disposition", "attachment; filename=\"pedidos-" + LocalDate.now() + ".csv\"");
        response.getWriter().write(csv);
    }

    @GetMapping("/{id}")
    public PedidoDetalle obtenerUno(@PathVariable String id) {
        return pedidoService.obtenerUno(id);
    }

    @GetMapping("/{id}/instrucciones-pago")
    public PedidoInstruccionesPagoDto obtenerInstruccionesPago(@PathVariable String id) {
        return pedidoService.obtenerInstruccionesPago(id);
    }

    @PutMapping("/{id}/confirmar-pago")
    public PedidoOperacionesDto.ConfirmarPagoResult confirmarPago(@PathVariable String id) {
        return pedidoService.confirmarPago(id);
    }

    @PutMapping("/{id}/cancelar")
    public PedidoOperacionesDto.CancelarPedidoResponse cancelar(@PathVariable String id) {
        return pedidoService.cancelar(id);
    }
}
